docs - add examples for dicts and nested dicts in main/argument_specs.yml (#79994)
ansible/ansible: docs/docsite/rst/playbook_guide/playbooks_reuse_roles.rst

        * A short, one-line description of the entry point.
        * The ``short_description`` is displayed by ``ansible-doc -t role -l``.
        * The short description should always be a string and never a list.

    :description:

        * A longer description that may contain multiple lines.

>         * Every new list item is a new paragraph. You can have multiple sentences per paragraph.

Reviewer: ```suggestion
        * This can be a single string, or a list of strings. In case this is a list of strings, every list
           element is a new paragraph.
```
I would probably skip `You can have multiple sentences per paragraph.` here; I only added it in the example to emphasize that you don't have to put every sentence in one paragraph (because without this sentence the example did exactly that :) ).

        * A short, one-line description of the entry point.
        * The ``short_description`` is displayed by ``ansible-doc -t role -l``.

>         * The short description should always be a string and never a list.

Reviewer 2: ```suggestion
        * The short description should always be a string and never a list, and should not end in a period.
```

    :description:

        * A longer description that may contain multiple lines.
        * This can be a single string, or a list of strings. In case this is a list of strings, every list
           element is a new paragraph.

    :version_added:
    
        * The version of the role when the entrypoint was added.

>         * This is a string, and not a float, for example, ``version_added: '2.1'``.

Reviewer 3: While I would prefer if we don't accidentally recommend non-semver versioning by mentioning it works, this is a practical note.

Author: I'll happily change it then. :)

        :option-name:

            * The name of the option/argument.

        :description:

            * Detailed explanation of what this option does. It should be written in full sentences.

        :version_added:
        

>             * Only needed if this option was extended after the initial role release, in other words, this is greater than the top level 'version_added' field.

Reviewer: ```suggestion
            * Only needed if this option was added after the initial role/entry point release, in other words, this is greater than the top level ``version_added`` field.
```

            * This is a string, and not a float, for example, ``version_added: '2.1'``.

>             * In collections, this must be the collection version the option was added to. For example, ``version_added: 1.0.0``.

Reviewer 3: ~To be complete `version_added` or `version_added_collection` should be allowed imo.~* This hasn't been supported, and invalid keys will give a warning soon (and be sanitized out of the spec), so this should be coordinated with https://github.com/ansible/ansible/pull/79657/files#r1063606171.

\* https://github.com/ansible/ansible/pull/79657/files#r1063606171 is matching this PR now, so you can disregard my comment.

Reviewer: `version_added` has been supported by antsibull-docs for role argument specs basically since the beginning.

`version_added_collection` hasn't been supported by antsibull-docs since it didn't make any sense so far; after all you cannot have docs fragments for role argument specs (so new options cannot come from other collections).

Reviewer: (Actually for options antsibull-docs probably allows `version_added_collection` since it uses the same validation as for plugins.)

Author: So, I guess I can consider this resolved?

Reviewer: I would say so.

        :required:

            * Only needed if ``true``.
            * If missing, the option is not required.

        :default:

            * If ``required`` is false/missing, ``default`` may be specified (assumed 'null' if missing).
            * Ensure that the default value in the docs matches the default value in the code. The actual
              default for the role variable will always come from ``defaults/main.yml``.
            * The default field must not be listed as part of the description, unless it requires additional information or conditions.
            * If the option is a boolean value, you should use `true/false` if you want to be compatible with `ansible-lint`.

        :choices:

            * List of option values.
            * Should be absent if empty.

        :elements:

            * Specifies the data type for list elements when type is ``list``.

        :options:

            * If this option takes a dict or list of dicts, you can define the structure here.

Sample specification
--------------------

.. code-block:: yaml

  # roles/myapp/meta/argument_specs.yml
  ---
  argument_specs:
    # roles/myapp/tasks/main.yml entry point
    main:

>       short_description:
>         - The main entry point for the myapp role.

Reviewer: The short description should always be a string and never a list:
```suggestion
      short_description: The main entry point for the myapp role.
```
(antsibull-docs will scream at you if you put a list here.)

Author: I'll change that. I don't like tools screaming at me... :)

Author: I will also put that hint in the description above.

      description:
        - This is the main entrypoint for the C(myapp) role.

>         - Here we can describe what this entrypoint does in lenghty words.

Reviewer: ```suggestion
        - Here we can describe what this entrypoint does in lenghty words.
        - Every new list item is a new paragraph. You can have multiple sentences
          per paragraph.
```

Author: I will also put that hint in the description above. So in case somebody only reads the explanation of the `description` field and does not read this example.

            * Ensure that the default value in the docs matches the default value in the code. The actual
              default for the role variable will always come from ``defaults/main.yml``.
            * The default field must not be listed as part of the description, unless it requires additional information or conditions.
            * If the option is a boolean value, you should use `true/false` if you want to be compatible with `ansible-lint`.

        :choices:

            * List of option values.
            * Should be absent if empty.

        :elements:

            * Specifies the data type for list elements when type is ``list``.

        :options:

            * If this option takes a dict or list of dicts, you can define the structure here.

Sample specification
--------------------

.. code-block:: yaml

  # roles/myapp/meta/argument_specs.yml
  ---
  argument_specs:
    # roles/myapp/tasks/main.yml entry point
    main:
      short_description: The main entry point for the myapp role.
      description:
        - This is the main entrypoint for the C(myapp) role.
        - Here we can describe what this entrypoint does in lenghty words.

>         - Here we can describe what this entrypoint does in lenghty words.

Reviewer: I think this line is now here twice.

        :option-name:

            * The name of the option/argument.

        :description:

            * Detailed explanation of what this option does. It should be written in full sentences.

        :version_added:
        

>             * Only needed if this option was added after the initial role/entry point release, in other words, this is greater than the top level ``version_added`` field.

Reviewer 2: ```suggestion
            * Only needed if this option was added after the initial role/entry point release. In other words, this is greater than the top level ``version_added`` field.
```

        :required:

            * Only needed if ``true``.
            * If missing, the option is not required.

        :default:

            * If ``required`` is false/missing, ``default`` may be specified (assumed 'null' if missing).
            * Ensure that the default value in the docs matches the default value in the code. The actual
              default for the role variable will always come from ``defaults/main.yml``.
            * The default field must not be listed as part of the description, unless it requires additional information or conditions.
            * If the option is a boolean value, you should use `true/false` if you want to be compatible with `ansible-lint`.

        :choices:

            * List of option values.
            * Should be absent if empty.

        :elements:

            * Specifies the data type for list elements when type is ``list``.

        :options:

            * If this option takes a dict or list of dicts, you can define the structure here.

Sample specification
--------------------

.. code-block:: yaml

  # roles/myapp/meta/argument_specs.yml
  ---
  argument_specs:
    # roles/myapp/tasks/main.yml entry point
    main:

>       short_description: The main entry point for the myapp role.

Reviewer 2: ```suggestion
      short_description: Main entry point for the myapp role
```

Reviewer 2: Short description also becomes part of the title for the role page in documentation. So it should be a phrase, not a sentence, and not have a period at the end.

      description:
        - This is the main entrypoint for the C(myapp) role.

>         - Here we can describe what this entrypoint does in lenghty words.

Reviewer 3: ```suggestion
        - Here we can describe what this entrypoint does in lengthy words.
```

Author: I'll change it as suggested.

.. code-block:: yaml

  # roles/myapp/meta/argument_specs.yml
  ---
  argument_specs:
    # roles/myapp/tasks/main.yml entry point
    main:

>       short_description: The main entry point for the myapp role.

Reviewer: ```suggestion
      short_description: The main entry point for the myapp role.
      description:
        - This is the main entrypoint for the C(myapp) role.
        - Here we can describe what this entrypoint does.
```

      options:
        myapp_int:
          type: "int"
          required: false
          default: 42

>           description: "The integer value, defaulting to 42."

Reviewer: ```suggestion
          description:
            - "The integer value, defaulting to 42."
            - This is a second paragraph.
```

        myapp_str:
          type: "str"
          required: true
          description: "The string value"

        myapp_list:
          type: "list"
          elements: "str"
          required: true

>           description: "A list of string values."

Reviewer: ```suggestion
          description: "A list of string values."
          version_added: 1.3.0
```

              description: "Time to prepare a dish in minutes."

        myapp_dict:
          type: "dict"
          required: false
          default:
            myapp_host: "bar.foo"
            myapp_exclude_host: true
            myapp_path: "/etc/myapp"
          description: "A dict with a defined structure and default values."
          options:
            myapp_host:
              type: "str"
              choices:
                - "foo.bar"
                - "bar.foo"
                - "ansible.foo.bar"
              required: true
              description: "A string value with a limited list of allowed choices."
            
            myapp_exclude_host:
              type: "bool"
              required: true
              description: "A boolean value."
            
            myapp_path:
              type: "path"
              required: true
              description: "A path value."
            
            original_name:
              type: list
              elements: "str"
              required: false
              description: "An optional list of string values."

    # roles/myapp/tasks/alternate.yml entry point
    alternate:

>       short_description: The alternate entry point for the myapp role.

Reviewer: ```suggestion
      short_description: The alternate entry point for the myapp role.
      version_added: 1.2.0
```

Author: There is only a `version_added` for options, it's ignored at the entrypoint level.

Reviewer: antsibull-docs has been supporting it since role support was added, so it's not ignored there.
